### DISCUSSION TITLE
Is that normal?

### DISCUSSION THREAD

**Discussion by jokoon** [foxssake/netfox#177] [General]
I tested netfox locally, and I can see some lag.

https://i.imgur.com/xpdEs3F.mp4

is that a matter of configuration?

**Comment by TheYellowArchitect** [comment]
maybe its the FPS, its under the physics tick rate. The FPS is heavy because of graphics or other things I assume because without any connections the game tanks at 200 FPS.

**Reply by elementbound** [reply]
To clarify, the network logic runs at 24 fps in Forest Brawl, and is not dependent on the physics tick rate ( unless configured so ).

**Comment by elementbound** [comment]
Thanks for reaching out @jokoon, let me get back to you on this one. My initial guess is the tickrate - the network loop runs at 24fps, so even on local, there could be an 1/24th second of latency.

**Reply by jokoon** [reply]
That would a 41 ms lag.

I did not measure, doesn't the video show more than double than that?

**Reply by elementbound** [reply]
Checked locally and you're right, it's definitely more than that. I've ran some artificial latency configs on my end and your video feels more like 200ms RTT. What's interesting is that the left window seemingly shows more latency than the right one. Is the right window running in host? 

Also, did you modify the display offset under Project Settings > Netfox > Rollback by chance?

**Reply by robinreversi** [reply]
Giving this a bump, did anyone figure out what's going on? I've noticed this too both in the demo and in my own project.

**Reply by elementbound** [reply]
Hi @robinreversi, nothing specific yet, unfortunately. What HW and OS are you running on?

**Reply by albertok** [reply]
@elementbound 
I've seen it also - what's happening is if a process frame takes longer than a tick's theoretical allocated time it falls behind by a few ticks and doesn't correct till a hard time sync.

You can see it really clearly if you add a line like this to time-display.gd under the Remote Time line and do something that can take a long time like a shader loading or a NavAgent calculating complex avoidance.

```swift
text += "\nLocal Time: %.2f at tick #%d, time diff: %.2f, tick diff: #%d" % [NetworkTime.time, NetworkTime.tick, NetworkTime.remote_time - NetworkTime.time, NetworkTime.remote_tick - NetworkTime.tick]
```

It's a tricky one to fix and I'm wondering if it would be more sane to work exclusively with the physics_process times and keep a consistent tickrate based on that.

![Screenshot_20240527_122618](https://github.com/foxssake/netfox/assets/5311294/5895864d-c21f-4ce6-a3cc-bf1e4f9567a5)

**Reply by robinreversi** [reply]
Hey @elementbound, running on MacOS Sonoma 14.3 with the M2 chip. I tried adding the logs albertok mentioned, and I'm also seeing the clients fall behind on ticks.

@albertok, sorry a bit new to networking, could you explain a bit more what you mean by "it would be more sane to work exclusively with the physics_process times and keep a consistent tickrate based on that."?

**Reply by albertok** [reply]
@robinreversi 

It's not so much a networking thing but a Godot engine internals one. They kind of mention it [here](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-private-method-process)

Basically running the ticks based on _process works ok until you get a frame drop. _process is meant to be called each frame rendered and we are calculating ticks from that. The problem is that it seems like it's not the most reliable of clocks to base our calculations from.

_physics_process is [supposed to be more reliable](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-private-method-physics-process). There is an option to sync netfox to physics but your tick rate will be forced to run at whatever the physics engine does, meaning if its running at 60 fps, that's the number of ticks fired which can overwhelm the rpcs and rollbacks being performed.

We likely need a middle ground where we can run a tick based on physics_frames but not every physics frame. ie -every second physics_frame at 60 fps would be 30.

**Reply by albertok** [reply]
Thinking on it more - would it be best to not rely on the game loops at all and leverage the system clock directly?

**Reply by elementbound** [reply]
Thanks both! 

---

@albertok I've assumed `_process` is called basically as often as possible, and the time between calls is provided as `delta`, so I didn't expect frame drops to be possible. Similar assumption for `_physics_process` - the engine runs on a single thread, and in each step, `_physics_process` is ran as many times as it needs to catch up with its fixed timestep. 

I wonder if there's any documentation about this, since all my assumptions are based on how Unity does [time](https://docs.unity3d.com/Manual/class-TimeManager.html). 

From what I've found, [delta smoothing](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-application-run-delta-smoothing) may or may not mess things up if V-Sync is enabled. 

As for solutions, I'm leaning towards using `_process` with the OS clock - it provides more direct access compared to `_physics_process` imo. Unless you know about some way to skip both `_process` and `_physics_process` :slightly_smiling_face:  I'll submit a PR with the change; if you happen to have the time, I'd love to hear about your results testing it as well.

Additionally, I expect implementing #176 should fix this properly.

---

@robinreversi for now, you could try enabling the *Netfox/Time/Sync to physics* setting, changing your physics tickrate to whatever the network tickrate was, and see if the issue persists. Let me know if this solves the issue!

**Reply by albertok** [reply]
Did a bit of testing this morning.

30 hz physics / sync to physics on / vsync on
30 hz physics / sync to physics on / vsync off 
30 hz physics / sync to physics off / vsync on
30 hz physics / sync to physics off / vsync off

All tests were performed in windowed mode.

They all seem to drift eventually from the server but 30 hz physics / sync to physics on / vsync off seemed to be the most consistent.

Running physics at 30 fps looks really bad, you get a stop motion effect. I didn't realise Godot wouldn't interpolate - [its being worked on though. ](https://github.com/godotengine/godot/pull/92391)

@elementbound Given the above your solution makes sense to me. Pin ticks to a reliable clock like the system one and ditch any reliance on loop deltas.

**Reply by elementbound** [reply]
Thanks for the testing @albertok! Here's the draft PR for using the OS clock, let me know what you think and if it still drifts as strongly: #225

**Reply by robinreversi** [reply]
Tried enabling sync to physics but ran into some other bugs, will let you know once I get these sorted out. Thanks both for the help and the quick turnaround time!

**Reply by elementbound** [reply]
@robinreversi just updated #225, hopefully it should fix the time drifting even without sync to physics. Feel free to test it as-is and let me know what you find! If you feel like, feel free to leave a review on the PR as well, in case you spot something :slightly_smiling_face:

**Reply by robinreversi** [reply]
Just gave it a try, seems like it worked, thanks!!

**Comment by jokoon** [comment]
Yes I think the right one is the host, but I am not sure... one the other
is the host.

Note: it was done on a slow laptop, i5 without GPU, so that might be the
issue.

No I did not modify that offset.

**Comment by elementbound** [comment]
@jokoon @robinreversi the fix is merged to main! Please feel free to open a new discussion / issue if you encounter any problems.

Thanks @albertok for the investigation and help with the PR! 

Closing topic.
